pinot-core/src/main/java/org/apache/pinot/core/query/selection/SelectionOperatorService.java

```diff
@@ -60,74 +65,255 @@ public class SelectionOperatorService {
+    List<Object[]> mergedRows = nWayMergeDataTables(dataTableList);
+    return new ResultTable(_dataSchema, mergedRows);
+  }
+
+  /**
+   * Merge sorted dataTables using N-way merge
+   * @param dataTables sorted dataTables
+   * @return sorted rows
+   */
+  private List<Object[]> nWayMergeDataTables(List<DataTable> dataTables) {
+    Comparator<Object[]> comparator = OrderByComparatorFactory.getComparator(_queryContext.getOrderByExpressions(),
+        _queryContext.isNullHandlingEnabled());
+    Comparator<MergeItem> mergeItemComparator = (MergeItem o1, MergeItem o2) -> comparator.compare(o1._row, o2._row);
+    PriorityQueue<MergeItem> mergeSortRows =
+        new PriorityQueue<>(Math.min(_numRowsToKeep, SelectionOperatorUtils.MAX_ROW_HOLDER_INITIAL_CAPACITY),
+            mergeItemComparator);
+
+    // populate pq
+    int numDataTables = dataTables.size();
+    int[] nextRowIds = new int[numDataTables];
+    int[] numRows = new int[numDataTables];
+    RoaringBitmap[][] dataTableNullBitmaps = getdataTableNullBitmaps(dataTables);
+    for (int i = 0; i < numDataTables; i++) {
+      DataTable dataTable = dataTables.get(i);
+      numRows[i] = dataTable.getNumberOfRows();
+      if (numRows[i] > 0) {
+        Object[] row = getDataTableRow(dataTable, 0, dataTableNullBitmaps[i]);
+        MergeItem mergeItem = new MergeItem(row, i);
+        mergeSortRows.add(mergeItem);
+        nextRowIds[i] = 1;
+      }
+    }
+
+    // merge
+    List<Object[]> resultRows = new ArrayList<>();
+    DataSchema.ColumnDataType[] columnDataTypes = _dataSchema.getColumnDataTypes();
+    int numColumns = columnDataTypes.length;
+    int offsetCounter = _offset;
+    while (resultRows.size() < _limit && !mergeSortRows.isEmpty()) {
+      MergeItem item = mergeSortRows.poll();
+      if (offsetCounter > 0) {
+        offsetCounter--;
+      } else {
+        Object[] row = item._row;
+        Object[] resultRow = formatRow(numColumns, row, columnDataTypes);
+        resultRows.add(resultRow);
+      }
+      int dataTableId = item._dataTableId;
+      int nextRowId = nextRowIds[dataTableId]++;
+      if (nextRowId >= numRows[dataTableId]) {
+        continue;
+      }
+      DataTable dataTable = dataTables.get(dataTableId);
+      Object[] row = getDataTableRow(dataTable, nextRowId, dataTableNullBitmaps[dataTableId]);
+      MergeItem mergeItem = new MergeItem(row, dataTableId);
+      mergeSortRows.add(mergeItem);
+    }
+    return resultRows;
+  }
+
+  private List<Object[]> processSingleDataTable(DataTable dataTable) {
+    List<Object[]> resultRows = new ArrayList<>();
+    int offsetCounter = _offset;
+    DataSchema.ColumnDataType[] columnDataTypes = _dataSchema.getColumnDataTypes();
+    int numColumns = _dataSchema.size();
+    int numRows = dataTable.getNumberOfRows();
+
+    if (_queryContext.isNullHandlingEnabled()) {
+      RoaringBitmap[] nullBitmaps = getNullBitmap(dataTable);
+      for (int rowId = 0; rowId < numRows; rowId++) {
```

Comment:
For single data table case, we can directly read the final rows:
- if numRows <= _offset: empty
- if numRows <= _offset + _limit: read from _offset to numRows
- if numRows > _offset + _limit: read from _offset to _offset + _limit